### ISSUE TITLE
Separate database per each process

### ISSUE THREAD

**Issue by dxops** [facile-it/paraunit-testcase#11]
Consider to use separate database for each process, that will allow us to avoid I/O issues

**Comment by Jean85** [comment]
How would you suggest to do this? The testcase doesn't know anything about an external DB, it's all behind Doctrine...

We considered this approach when we first started working on Paraunit, but the setup cost of each DB was too expensive for our project: the schema was too big and we had too much fixtures; so we ended up with this package. Also, spinning up multiple MySQL instance is more resource intensive in comparison to the current approach.

**Comment by dxops** [comment]
At this moment during CI we do application installation and clone databases and source and run tests in parallel. Assuming all tests are isolated, we have to switch DB only.

Good option is to get DB name from environment, like https://github.com/liuggio/fastest/blob/master/adapters/Doctrine/DBAL/ConnectionFactory.php#L25

No need to set up a lot of MySQL instances, few databases on one instance works faster that Paraunit now (mentioned small regression between 0.7 and 0.8 BTW). Configuration matters here - connections, buffers, e.g.

**Comment by Jean85** [comment]
Ok, so you're suggesting an approach similar to fastest. But in this case I would need to make Paraunit and this testcase to work together, since the runner would have to tell to the testcase something like an ID to know which database it should use.

This would mean merging this package into `facile-it/paraunit` to make this feasible.

**Comment by dxops** [comment]
We can do indirect dependency, just tell the current process ID in Paraunit and use it to generate DB name in Paraunit TestCase

**Comment by Jean85** [comment]
Yeah but this would mean that this testcase could work only with Paraunit, so it would have to be merged, or require (and no more just suggest) Paraunit.

Also: are you suggesting to generate the DB inside the testcase on its own? It would be incredibly slow! The DB should be ready before running the tests IMHO.

**Comment by dxops** [comment]
> Also: are you suggesting to generate the DB inside the testcase on its own? It would be incredibly slow! The DB should be ready before running the tests IMHO.

Agree, database copies should be created by CI server.

**Comment by Jean85** [comment]
I've done some tests, and it's even better if you create one DB and clone it afterwards with a `mysqldump`

**Comment by dxops** [comment]
Yes, we do initial installation and clone existing database atm.

For Postgres it's even faster - `CREATE DATABASE ${dbname}_$i WITH TEMPLATE ${dbname};`
